[BLOCKS DL] Retry connect to local node each 120s instead of every block

turbolay opened the pull request:
Trying to connect to a local node with no local node running currently loses 1s every block download due to the timeout for `ConnectAsync` to raise `SocketError`
Instead, this PR creates a runtime parameter (readonly, default 120s) to wait some time before retrying.
It also reports this behaviour as LogInfo (instead of LogTrace).

Other behaviours may also be considered, for example if Knots is enabled maybe we should retry every time? Also it could be better to add an UI cue for the user to know without looking at logs, especially if Knots is enabled.

turbolay commented on WalletWasabi/Wallets/P2pBlockProvider.cs:
```diff
@@ -74,12 +76,14 @@ public async Task<Block> GetBlockAsync(uint256 hash, CancellationToken cancellat
 				{
-					// Try to get block information from local running Core node first.
-					block = await TryDownloadBlockFromLocalNodeAsync(hash, cancellationToken).ConfigureAwait(false);
+					if(!LastFailConnectLocalNode.IsRunning || LastFailConnectLocalNode.Elapsed.TotalSeconds >= RuntimeParams.Instance.RetryConnectLocalNodeSeconds) { 
```
> `GetBlockAsync` method seems like it can be called multiple times (why not?). Or am I missing something?

The point is to wait for N seconds before retrying, even during a new call. 
But you brought a good point, I have to add a lock otherwise the logs will look bad in cases of concurrent calls

turbolay commented on WalletWasabi/Helpers/RuntimeParams.cs:
```diff
@@ -15,10 +15,13 @@ public class RuntimeParams
 	{
 	}
-
+	
```
I use VS but since I began to contribute to WW I have ton of similar issues with it... No auto-formatting, no warnings/erros highlighting, Ctrl+k not working etc...
I'll try to reinstall

turbolay commented:
I realized that when there are few nodes connected chat was spammed by the LogInfo message so it's now displayed only once (except if a node is OFF then ON then OFF again, message will be displayed again)

I also added a Semaphore. It was a bit tricky because we should not prohibit parallelization of block download from local node, only the connection should be locked. As the behaviour of the function is not trivial with various try/catch etc I preferred not to refactor, but if my Semaphore implementation is not clear I can split in 2 functions, Connect to and dl from local node.

lontivero reviewed (commented):
I'm not a fan of the CodeScene robot (i don't really like our robots) but this time i can only agree that the code is already extremely shitty and this work makes it even worse.

turbolay commented:
So I will refactor tomorrow with proper AsyncLock

turbolay commented:
I did some refactoring, it ain't much but I think it's way better now that code is splitted in several functions.
I tried to follow guidelines and current codebase style as much as i could but please give me feedback if some parts (especifically the if elseif else) diverge.

yahiheb commented:
Reopened as requested by @turbolay

turbolay commented on WalletWasabi/Wallets/P2pBlockProvider.cs:
```diff
@@ -74,14 +79,35 @@ public async Task<Block> GetBlockAsync(uint256 hash, CancellationToken cancellat
+					{
+						using (await ConnectLocalNodeLock.LockAsync(cancellationToken).ConfigureAwait(false))
+						{
+							await TryConnectToLocalNodeAsync(cancellationToken).ConfigureAwait(false);
```
New function to separate the connection code from `TryDownloadBlockFromLocalNodeAsync`

turbolay commented:
@molnard ready for review.
I changed a bit the concept: It will try each 1s then 4, 16, 64, and then 120s
I also added comments on the PR to help review as it's almost only refactoring.

molnard reviewed (changes requested):
I am sure this is working but the code is messy and very hard to review. - however, this should be a simple logic. The complexity added historically - I think it would be worth the time now to clean it. 
For example, Local node-related code could be moved to a different class. 
Maybe others as well?
